packages/gatsby-source-contentful/src/extend-node-type.js

```diff
@@ -542,18 +541,83 @@ const fluidNodeType = ({ name, getTracedSVG }) => {
   }
 }
 
-exports.extendNodeType = ({ type, store }) => {
+exports.extendNodeType = ({ type, store, cache }) => {
   if (type.name.match(/contentful.*RichTextNode/)) {
     return {
       nodeType: {
         type: GraphQLString,
-        deprecationReason: `This field is deprecated, please use 'json' instead.`,
+        deprecationReason: `This field is deprecated, please use 'raw' instead. @todo add link to migration steps.`,
       },
       json: {
-        type: GraphQLJSON,
-        resolve: (source, fieldArgs) => {
-          const contentJSON = JSON.parse(source.internal.content)
-          return contentJSON
+        type: GraphQLString,
+        deprecationReason: `This field is deprecated, please use 'raw' instead. @todo add link to migration steps.`,
+      },
+      references: {
+        type: [`ContentfulReference`],
+        async resolve(source, args, context, info) {
+          const parent = await context.nodeModel.findRootNodeAncestor(source)
+
+          const rawReferences = { Entry: [], Asset: [] }
+
+          // Locate all Contentful Links within the rich text data
+          const traverse = obj => {
+            for (let k in obj) {
+              const value = obj[k]
+              if (value && value.sys && value.sys.type === `Link`) {
+                rawReferences[value.sys.linkType].push(value.sys.contentful_id)
+              } else if (value && typeof value === `object`) {
+                traverse(value)
+              }
+            }
+          }
+
+          traverse(JSON.parse(source.raw))
+
+          if (!rawReferences.Entry.length && !rawReferences.Asset.length) {
+            return null
+          }
+
+          // Query for referenced nodes
+          rawReferences.Entry = [...new Set(rawReferences.Entry)]
+          rawReferences.Asset = [...new Set(rawReferences.Asset)]
```

Comment:
Maybe create `Entry` and `Asset` as `Set` in the first place (and update `traverse` accordingly):

```js
const rawReferences = { Entry: new Set(), Asset: new Set() }
```

Will save several CPU ticks :)